File: urbanairship-automation/src/main/java/com/urbanairship/automation/deferred/StateOverrides.java
/* Copyright Airship and Contributors */

package com.urbanairship.automation.deferred;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.annotation.VisibleForTesting;

import com.urbanairship.BuildConfig;
import com.urbanairship.UAirship;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonSerializable;
import com.urbanairship.json.JsonValue;
import com.urbanairship.push.PushManager;

import java.util.Locale;

/**
 * A model defining state overrides.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class StateOverrides implements JsonSerializable {

    private static final String STATE_APP_VERSION_KEY = "app_version";
    private static final String STATE_SDK_VERSION_KEY = "sdk_version";
    private static final String STATE_NOTIFICATION_OPT_IN_KEY = "notification_opt_in";
    private static final String STATE_LOCALE_LANGUAGE_KEY = "locale_language";
    private static final String STATE_LOCALE_COUNTRY_KEY = "locale_country";

    private final int appVersion;
    private final String sdkVersion;
    private final boolean notificationOptIn;
    private final String localeLanguage;
    private final String localeCountry;

    /**
     * Default state overrides constructor.
     */
    @VisibleForTesting
    StateOverrides(int appVersion, @NonNull String sdkVersion, boolean notificationOptIn, @NonNull Locale locale) {
        this.appVersion = appVersion;
        this.sdkVersion = sdkVersion;
        this.notificationOptIn = notificationOptIn;
        localeLanguage = locale.getLanguage();
        localeCountry = locale.getCountry();
    }

    /**
     * Creates a state overrides.
     *
     * @return An instance of StateOverrides.
     */
    @NonNull
    public static StateOverrides defaultOverrides() {
        PushManager pushManager = UAirship.shared().getPushManager();
        Locale locale = UAirship.shared().getLocale();
        return new StateOverrides(BuildConfig.VERSION_CODE, BuildConfig.SDK_VERSION, pushManager.isOptIn(), locale);
    }

    @NonNull
    @Override
    public JsonValue toJsonValue() {
        return JsonMap.newBuilder()
                .put(STATE_APP_VERSION_KEY, appVersion)
                .put(STATE_SDK_VERSION_KEY, sdkVersion)
                .put(STATE_NOTIFICATION_OPT_IN_KEY, notificationOptIn)
                .put(STATE_LOCALE_LANGUAGE_KEY, localeLanguage)
                .put(STATE_LOCALE_COUNTRY_KEY, localeCountry)
                .build()
                .toJsonValue();
    }
}

File: urbanairship-core/src/main/java/com/urbanairship/deferred/StateOverrides.kt
package com.urbanairship.deferred

import com.urbanairship.json.JsonSerializable
import com.urbanairship.json.JsonValue
import com.urbanairship.json.jsonMapOf
import java.util.Locale

internal data class StateOverrides(
    val appVersionName: String,
    val sdkVersion: String,
    val notificationOptIn: Boolean,
    val locale: Locale?
) : JsonSerializable {
    companion object {
        private const val KEY_APP_VERSION = "app_version"
        private const val KEY_SDK_VERSION = "sdk_version"
        private const val KEY_NOTIFICATION_OPT_IN = "notification_opt_in"
        private const val KEY_LOCALE_LANGUAGE = "locale_language"
        private const val KEY_LOCALE_COUNTRY = "locale_country"
    }

    constructor(request: DeferredRequest) : this(
        appVersionName = request.appVersionName,
        sdkVersion = request.sdkVersion,
        notificationOptIn = request.notificationOptIn,
        locale = request.locale
    )

    override fun toJsonValue(): JsonValue {
        return jsonMapOf(
            KEY_APP_VERSION to appVersionName,
            KEY_SDK_VERSION to sdkVersion,
            KEY_NOTIFICATION_OPT_IN to notificationOptIn,
            KEY_LOCALE_COUNTRY to locale?.country,
            KEY_LOCALE_LANGUAGE to locale?.language
        ).toJsonValue()
    }
}
